libcache/cache.c

```diff
@@ -0,0 +1,696 @@
+		}
+	}
+
+	return position;
+}
+
+
+int cache_deinit(cachectx_t *cache)
+{
+	int i, j, ret = EOK, err;
+	uint64_t addr;
+	ssize_t written;
+	cacheline_t *linePtr;
+
+	if (cache == NULL) {
+		ret = -EINVAL;
+	}
+	else {
+		for (i = 0; i < cache->numSets && ret != -EIO; ++i) {
+			for (j = 0; j < LIBCACHE_NUM_WAYS; ++j) {
+				linePtr = &(cache->sets[i].lines[j]);
+
+				if (IS_VALID(linePtr->flags)) {
+					if (IS_DIRTY(linePtr->flags)) {
+						addr = cache_computeAddr(cache, linePtr->tag, i);
+						written = cache_flushLine(cache, linePtr, addr);
+
+						if (written < (ssize_t)cache->lineSize) {
+							ret = -EIO;
+							break;
+						}
+					}
+
+					free(linePtr->data);
+					CLEAR_VALID(linePtr->flags);
+				}
+			}
+		}
+
+		if (ret != -EIO) {
+			err = resourceDestroy(cache->lock);
+			if (err < 0) {
+				ret = err;
+			}
+			else {
+				free(cache->sets);
+				free(cache);
+			}
+		}
+	}
+
+	return ret;
+}
+
+
+static int cache_compareTags(const void *lhs, const void *rhs)
+{
+	int ret = 0;
+
+	const cacheline_t **lhsLine = (const cacheline_t **)lhs;
+	const cacheline_t **rhsLine = (const cacheline_t **)rhs;
+
+	if (*lhsLine == NULL && *rhsLine == NULL) {
+		ret = 0;
+	}
+	else if (*lhsLine == NULL && *rhsLine != NULL) {
+		ret = -1;
+	}
+	else if (*lhsLine != NULL && *rhsLine == NULL) {
+		ret = 1;
+	}
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+static ssize_t cache_executePolicy(cachectx_t *cache, cacheline_t *linePtr, uint64_t addr, int policy)
+{
+	ssize_t written = cache->lineSize;
+
+	if (policy == 1) {
+		written = cache_flushLine(cache, linePtr, addr);
+		if (written <= 0) {
+			written = -EIO;
+		}
+	}
+
+	return written;
+}
+
+
+static cacheline_t *cache_allocateLine(cachectx_t *cache, const uint64_t setIndex, const uint64_t tag)
+{
+	int i, lineIndex;
+	uint64_t addr = 0;
+	cacheline_t *temp = NULL, *linePtr = NULL;
+	unsigned char flags = 0;
+	void *data;
+	ssize_t written = cache->lineSize;
+	cacheset_t *setPtr = &cache->sets[setIndex];
+
+	if (setPtr->count < LIBCACHE_NUM_WAYS) {
+		setPtr->count++;
+
+		for (i = 0; i < setPtr->count; ++i) {
+			if (!IS_VALID(setPtr->lines[i].flags)) {
+				lineIndex = i;
+				break;
+			}
+		}
+
+		temp = &(setPtr->lines[lineIndex]);
+
+		for (i = 0; i < setPtr->count; ++i) {
+			if (setPtr->tags[i] == NULL) {
+				setPtr->tags[i] = temp;
+				break;
+			}
+		}
+
+		LIST_ADD(&setPtr->timestamps, temp);
+	}
+	else {
+		temp = setPtr->timestamps;
+
+		if (IS_DIRTY(temp->flags)) {
+			addr = cache_computeAddr(cache, temp->tag, setIndex);
+			written = cache_flushLine(cache, temp, addr);
+		}
+
+		if (written == (ssize_t)cache->lineSize) {
+			LIST_REMOVE(&setPtr->timestamps, temp);
+			free(temp->data);
+
+			for (i = 0; i < setPtr->count; ++i) {
+				if (IS_VALID(setPtr->lines[i].flags) && (setPtr->lines[i].tag == temp->tag)) {
```

Comment:
IS_VALID check should be redundant - all lines should be valid in this case

Reply:
You're right. I'll remove it.